### PR TITLE
add early stopping for GluonTS model during training

### PR DESCRIPTION
**Pull request by huibinshen**
*Issue #, if available:*

Enable training early stopping for GluonTS model.

Local tests passed.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1795-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1795/1/index.html

**Comment by gradientsky** [comment]
please add unit tests for early stopping logic.

**Review by canerturkmen — COMMENTED** [review]
I checked the GluonTS source. I remember we used to have early stopping by default but now I couldn't find it. In any case, why do we want to enable it for autogluon (as GluonTS has moved away and to learning rate scheduling by default) and if so, why do we want to enable it by default without giving the user the choice to do it or even adjust the patience?

Maybe for now should we just keep the callback, not add it to models, and add a unit test?

File: timeseries/src/autogluon/timeseries/models/gluonts/callback.py [hunk 1]
```diff
@@ -37,3 +37,26 @@ def on_epoch_end(
...
+        self.best_round = 0
+        self.best_loss = float('inf')
+
+    def on_validation_epoch_end(self, epoch_no, epoch_loss, **kwargs):
+        if epoch_loss < self.best_loss:
+            self.best_loss = epoch_loss
+            self.best_round = epoch_no
+            return True
+        else:
+            contniue = (epoch_no - self.best_round) < self.patience
```
Comment (Innixma) [inline timeseries/src/autogluon/timeseries/models/gluonts/callback.py:58]: nit: typo on continue

File: timeseries/src/autogluon/timeseries/models/gluonts/callback.py [hunk 2]
```diff
@@ -37,3 +37,26 @@ def on_epoch_end(
...
                 return False
         return True
+
+
+class EarlyStoppingCallback(Callback):
+    """GluonTS callback to early stop the training if the validation loss
+    is not improved for `patience' round. For the GluonTS models used in autogluon,
+    the loss is always minimized."""
+
+    def __init__(self, patience=10):
```
Comment (Innixma) [inline timeseries/src/autogluon/timeseries/models/gluonts/callback.py:47]: We need to be careful here, this is not a good patience value to use in all situations. This is a complex topic and I put a lot of effort into this in Tabular.

For example, I have Early stopping logic coded here: https://github.com/awslabs/autogluon/blob/master/core/src/autogluon/core/utils/early_stopping.py

In particular, AdaptiveES will be much better than naive patience based ES. But regardless, we should err on the side of caution and use high patience values rather than low ones, otherwise we risk crippling the models in situations where time isn't much of a concern.

Comment (huibinshen) [reply timeseries/src/autogluon/timeseries/models/gluonts/callback.py:48]: Thanks for the comments!

I am not very sure about 10 is not good for *all* situations but I agree that we need some data points to set this default value. Do you have some results to share about the adaptive early stopping?

I also implemented the adaptive early stopping logic for GluonTS model, there was quite some code duplicates and I am thinking of talking the `_update_patience` out of the class `AdaptiveES` so I could reuse it. Maybe you have better suggestion.

I also remove the default usage of early stopping for now. After we have more results for it, we could set the default option later based on that. Let me know what do you think of this option.

File: timeseries/src/autogluon/timeseries/models/gluonts/callback.py (hunk 1 above)
Comment (huibinshen) [reply timeseries/src/autogluon/timeseries/models/gluonts/callback.py]: Changed, thanks!

**Comment by huibinshen** [comment]
> I checked the GluonTS source. I remember we used to have early stopping by default but now I couldn't find it. In any case, why do we want to enable it for autogluon (as GluonTS has moved away and to learning rate scheduling by default) and if so, why do we want to enable it by default without giving the user the choice to do it or even adjust the patience?
> 
> Maybe for now should we just keep the callback, not add it to models, and add a unit test?
Thanks Caner for the comments!

I asked Lorenzo and Jasper about it, they all think that it can be worth to use it (looking back, they would have not removed it). But I think the default choice worth a study so I don't use it for now. After we have more data points, we could change the default option then. Let me know what do you think of this option.

**Comment by szha** [comment]
Job PR-1795-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1795/2/index.html

File: timeseries/src/autogluon/timeseries/models/gluonts/callback.py (hunk 2 above)
Comment (Innixma) [reply timeseries/src/autogluon/timeseries/models/gluonts/callback.py:48]: Sorry for confusion, when I said "all situations" I meant that in some situations it might not be good, not that it is always not good.

Regarding your implementation, to avoid code dupe why not directly use the AdaptiveES code? You can import it and use it in your callback to avoid code dupe. I do this for individual models, for example, having it as `self.es`: https://github.com/awslabs/autogluon/blob/master/tabular/src/autogluon/tabular/models/xgboost/callbacks.py#L27-L30

We probably shouldn't remove default usage of early stopping, just err on the side of caution and use a larger value. What value is best will be understood after we start doing benchmarking. Maybe 10 is fine but it depends on the model.

Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/models/gluonts/callback.py:48]: perhaps even 5 is fine if we have it on by default. if anything it's a little too cautious considering that we work with 50-100 epochs.

Comment (huibinshen) [reply timeseries/src/autogluon/timeseries/models/gluonts/callback.py]: Re the code duplicates, I followed your suggestions. For the default patience, I put it to 20 for now. We may change it later with more data points.

**Comment by szha** [comment]
Job PR-1795-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1795/3/index.html

**Comment by Innixma** [comment]
@canerturkmen I'll leave this PR to you, feel free to approve and merge if you wish.

**Comment by huibinshen** [comment]
Closing this and will create another one.
